### PR TITLE
Use selective bug cache for attachment count

### PR DESCRIPTION
**Pull request by cproensa**
Resubmit the changes from PR #654 as the main issue with long IN clauses was solved after refactoring export pages

This is a dramatic improvement in page load when there is a lot of attachments in the db.
See comment: https://github.com/mantisbt/mantisbt/pull/654#issuecomment-153838142

Additionally:
Added some changes to include this file count cache in the clearing procedures at export pages to save on memory.
Fixed some procedures for clearing caches that were wrong.

### CODE REVIEW

File: core/bugnote_api.php [hunk 1]
```diff
@@ -753,10 +753,38 @@ function bugnote_clear_cache( $p_bugnote_id = null ) {
...
+		}
 	}
-	$g_cache_bugnotes = array();
 
 	return true;
 }
+
+/**
+ * Clear the bugnotes related to a bug, or all bugs if no bug id specified.
+ * @param integer $p_bu_id Identifier to clear (optional).
```
Comment (vboctor) [inline core/bugnote_api.php]: Typo in parameter name.

File: core/bugnote_api.php [hunk 2]
```diff
@@ -753,10 +753,38 @@ function bugnote_clear_cache( $p_bugnote_id = null ) {
 
 	if( null === $p_bugnote_id ) {
 		$g_cache_bugnote = array();
+		$g_cache_bugnotes = array();
 	} else {
-		unset( $g_cache_bugnote[(int)$p_bugnote_id] );
+		if( isset( $g_cache_bugnote[(int)$p_bugnote_id] ) ) {
+			$t_note_obj = $g_cache_bugnote[(int)$p_bugnote_id];
+			bugnote_clear_bug_cache( $t_note_obj->bug_id );
+			unset( $g_cache_bugnote[(int)$p_bugnote_id] );
```
Comment (vboctor) [inline core/bugnote_api.php]: Would this still be set after clearing the bug and all associated notes?

File: core/file_api.php [hunk 3]
```diff
@@ -98,45 +98,81 @@ function file_get_display_name( $p_filename ) {
 }
 
 /**
- * Check the number of attachments a bug has (if any)
- * @param integer $p_bug_id A bug identifier.
- * @return integer
+ * Fills the cache with the attachement count from a list of bugs
+ * If the bug doesn't have attachments, cache its value as 0.
+ * @param array $p_bug_ids Array of bug ids
  */
```
Comment (vboctor) [inline core/file_api.php]: @ return void

File: core/file_api.php [hunk 4]
```diff
@@ -98,45 +98,81 @@ function file_get_display_name( $p_filename ) {
...
-	if( isset( $g_cache_file_count[$p_bug_id] ) ) {
-		return $g_cache_file_count[$p_bug_id];
+	if( empty( $p_bug_ids ) ) {
+		return;
+	}
+
+	$t_ids_to_sarch = array();
+	foreach( $p_bug_ids as $t_id ) {
+		$c_id = (int)$t_id;
+		$t_ids_to_sarch[$c_id] = $c_id;
```
Comment (vboctor) [inline core/file_api.php]: Typo in variable name.

File: core/filter_api.php [hunk 5]
```diff
@@ -2225,21 +2225,25 @@ function filter_get_bug_rows_query_clauses( array $p_filter, $p_project_id = nul
...
-		$b = $t_stats[$t_row['id']];
+		$t_bug_ids[] = (int)$t_row['id'];
 		if( array_key_exists( $t_row['id'], $t_stats ) ) {
 			$t_rows[] = bug_row_to_object( bug_cache_database_result( $t_row, $t_stats[$t_row['id']] ) );
 		} else {
 			$t_rows[] = bug_row_to_object( bug_cache_database_result( $t_row ) );
 		}
 	}
+	#cache the attachment count for bug list
+	file_bug_attachment_count_cache( $t_bug_ids );
```
Comment (vboctor) [inline core/filter_api.php]: Should we be doing this even if the caller has no intention of displaying this information (e.g. View Issues page without attachments column or soap API, etc)?  It feels that this function is not making assumptions about the callers intent.

File: core/bugnote_api.php (hunk 2 above)
Comment (cproensa) [inline core/bugnote_api.php]: what do you mean? having $t_note_obj as reference to the object, even after unsetting the index in the array?

File: core/bugnote_api.php [hunk 6]
```diff
@@ -753,10 +753,38 @@ function bugnote_clear_cache( $p_bugnote_id = null ) {
...
+ * @param integer $p_bug_id Identifier to clear (optional).
+ * @return boolean
+ * @access public
+ */
+function bugnote_clear_bug_cache( $p_bug_id = null ) {
+	global $g_cache_bugnotes, $g_cache_bugnote;
+
+	if( null === $p_bug_id ) {
+		$g_cache_bugnotes = array();
+		$g_cache_bugnote = array();
```
Comment (dregad) [inline core/bugnote_api.php]: I never noticed we had 2 distinct variables for bugnotes caching. It is not really clear what the difference between them are, and they are not formally declared at the top of bugnote api, only as global in the functions using them. Do we actually need 2 of them ?

Even though that's not directly related to this PR, I would suggest to declare them, and add a PHPDoc block to explain their purpose and differences.

File: core/bugnote_api.php (hunk 2 above)
Comment (dregad) [inline core/bugnote_api.php]: I think he means that in bugnote_clear_bug_cache(), all entries in $g_cache_bugnote for the corresponding bug_id are already unset (see [line 783](https://github.com/mantisbt/mantisbt/pull/941/files#diff-af4b0457065b17f59cdc20eec8005bc4R783))

File: core/filter_api.php (hunk 5 above)
Comment (dregad) [inline core/filter_api.php]: I think he means that in bugnote_clear_bug_cache(), all entries in $g_cache_bugnote for the corresponding bug_id are already unset (see [line 783](https://github.com/mantisbt/mantisbt/pull/941/files#diff-af4b0457065b17f59cdc20eec8005bc4R783))

File: core/bugnote_api.php (hunk 2 above)
Comment (vboctor) [inline core/bugnote_api.php]: My question is whether bugnote_clear_bug_cache() already clears the note from the $g_cache_bugnote, and then you are unsetting it again.  It feels to me that the unset line is redundant.

File: core/filter_api.php (hunk 5 above)
Comment (vboctor) [inline core/filter_api.php]: btw, "function is not making assumptions" -> "function is now making assumptions" :)

File: core/bugnote_api.php (hunk 6 above)
Comment (vboctor) [inline core/bugnote_api.php]: I would also rename them to more meaningful names.  For example:

- $g_cache_bugnotes_by_id
- $g_cache_bugnotes_by_bug_id

File: core/bugnote_api.php (hunk 2 above)
Comment (cproensa) [inline core/bugnote_api.php]: I got it now. you're right.
Apart from that... i'm not totally sure it the clearing of one cache should also clear the respective content in the other cache...

File: core/bugnote_api.php (hunk 6 above)
Comment (cproensa) [inline core/bugnote_api.php]: right, those names are self-explaining, way better than current ones.
however i would not rename at this moment that there is other PR rewritting the same api

Comment (dregad) [inline core/bugnote_api.php]: For follow-up https://www.mantisbt.org/bugs/view.php?id=21884

File: core/file_api.php [hunk 7]
```diff
@@ -98,45 +98,83 @@ function file_get_display_name( $p_filename ) {
...
+function file_bug_attachment_count_clear_cache( $p_bug_id = null ) {
+	global $g_cache_file_count;
+
+	if( null === $p_bug_id ) {
+		$g_cache_file_count = array();
+	} else {
+		unset( $g_cache_file_count[(int)$p_bug_id] );
+	}
+
+	return true;
```
Comment (vboctor) [inline core/file_api.php]: Why do we need a return value for this method?  It is always true.

File: core/filter_api.php [hunk 8]
```diff
@@ -2225,6 +2225,7 @@ function filter_get_bug_rows_query_clauses( array $p_filter, $p_project_id = nul
 
 /**
  *  Cache the filter results with bugnote stats for later use
+ *  also fills bug attachment count cache
```
Comment (vboctor) [inline core/filter_api.php]: Is this comment still valid?

File: core/file_api.php (hunk 7 above)
Comment (cproensa) [inline core/file_api.php]: The other clear cache methods used this scheme. 
No strong reason to return true, but a more detailed analysis on current usage, for the existing ones, should be done

**Comment by dregad** [comment]
Merged in d47f367fea8dfda1e5bce918419a1bda1dd0e10b

**Pull request closed without merging**